### PR TITLE
[#3897] Archive project via project settings by non admins

### PR DESCRIPTION
**Pull request by cbliard**
https://community.openproject.org/work_packages/3897

### CODE REVIEW

**Comment by cbliard** [comment]
Test `rspec ./spec/requests/api/v3/projects/update_resource_spec.rb:300` is failing. It is testing updating the `active` field of a project through the api.

Before implementing `archive_project` permission, it was testing `user.admin?`.
Now it tests `user.allowed_to?(:archive_project, project)`, and it always returns false if the project is archived. The way the API works, it first updates and persists the values, and then run the `Projects::ArchiveService` which fails without further notice.

That's why the test fails: the project is archived but not the subprojects.

~~I did not test but I have a hunch that it was possible for a normal user to archive a project through the API, bypassing the need to be an admin. I'll test it next time.~~ I was wrong. The permission was correctly checked and only an admin could change the active attribute and archive/unarchive a project through the API.

File: app/contracts/projects/unarchive_contract.rb
Comment (ulferts) [inline app/contracts/projects/unarchive_contract.rb:30]: Why having the `Unarchivine/Archive`-`Contracts` rely on the `::BaseContract`? That way, all of the checks in the `BaseContract` will be applied. E.g. if a required custom field is not set, the contract will fail. The user would not be able to solve that since the project is still archived so no custom fields can be set on the UI.

File: db/migrate/20221223084601_add_archive_project_permission_to_project_admin_role.rb
Comment (ulferts) [inline db/migrate/20221223084601_add_archive_project_permission_to_project_admin_role.rb]: I wouldn't add the permission automatically to existing roles unless we have a very good criteria of the roles we want to add the permission for. The amount of permission isn't a good criteria in my opinion since the seeded roles might have changed drastically. The way this is implemented now, the non member role might receive the permission.

In this case, I don't think that there is a solid criteria so I would leave as a manual task to add the permission to the correct roles.

File: config/initializers/permissions.rb
Comment (ulferts) [inline config/initializers/permissions.rb:42]: It might be a good idea to have `edit_project` as a dependency so that the user can reach the page (`projects/:id/settings/general`) where the button can be found. Of course, the user could achieve the same by visiting the project list but that is less intuitive. 

Probably more of a product question then it is a technical one.

File: app/contracts/projects/archive_contract.rb
Comment (ulferts) [inline app/contracts/projects/archive_contract.rb]: Why not combine the two loops?
```
unless model.descendants.all? { |subproject| user.allowed_to?(:archive_project, subproject) }
  errors.add :base, :archive_permission_missing_on_subprojects  
end
```

File: app/contracts/projects/base_contract.rb
Comment (ulferts) [inline app/contracts/projects/base_contract.rb:122]: This is another argument why `ArchiveContract` and `UnarchiveContract` should not depend on `BaseContract` (see below).

If any of those are called from here, within the `BaseContract`, the rest of the validations will be run twice at least. I currently don't understand why it seems to not be causing an infinite call stack.

File: app/models/users/project_role_cache.rb
Comment (ulferts) [inline app/models/users/project_role_cache.rb]: I'd rather call it "active state" then "activity" since the later is already used in different contexts in OP.

**Review by ulferts — CHANGES_REQUESTED** [review]
I haven't yet tried out the changes from the users point of view but just did a review of the code.

The change looks mostly solid. The thing I have the most questions about is the change to have the (un)archive contract inherit from the base contract but that might be just because I don't get it yet.

File: config/initializers/permissions.rb
Comment (cbliard) [reply config/initializers/permissions.rb:42]: Indeed. I asked the question two weeks ago to the product team: https://community.openproject.org/projects/openproject/work_packages/3897/activity#activity-83. I'm still waiting for a reply.

File: app/contracts/projects/archive_contract.rb
Comment (cbliard) [reply app/contracts/projects/archive_contract.rb]: Because it was easier to read it on multiple lines and without any `unless`. When I see `unless`, my brain has to convert it to `if not`.

On the other hand, the version you propose is more performant as it will stop on the first subproject where the user does not have the permission, whereas my version will scan for all subprojects.

Maybe I'd like to write this:

```ruby
if model.descendants.any? { |subproject| user.forbidden_to?(:archive_project, subproject) }
  errors.add :base, :archive_permission_missing_on_subprojects  
end
```

Can I add a `forbidden_to?` next to the `allowed_to?`?

And after digging deeper, it appears that `user#allowed_to?` accepts an array of projects, so it can be written like this instead:

```ruby
unless user.allowed_to?(:archive_project, model.descendants)
  errors.add :base, :archive_permission_missing_on_subprojects  
end
```

Seems better.

File: app/contracts/projects/unarchive_contract.rb
Comment (cbliard) [reply app/contracts/projects/unarchive_contract.rb:30]: Because it was not used: the model checks were skipped because of `validate_model?` being overridden to return `false`. That leave the `valid?()` method overriden in `ModelContract` only calling `readonly_attributes_unchanged`. And in the `Projects::ArchiveController`, the project instance being passed never has any modifications, so attributes were always left unchanged. So this is not necessary here.

When I tried to reuse the `Archive/Unarchive`-`Contract` in `Projects::UpdateContract`, at first it was not working because of the `readonly_attributes_unchanged` would block modifying anything if the project is archived (because all fields become readonly when the project is archived). I was looking to override `#readonly_attributes_unchanged` to do nothing, and then I realized that I was not using anything from `ModelContract`, so I changed to the parent class of `ModelContract`, which is `::BaseContract`

And then in the `Projects::UpdateService`, it calls `Projects::ArchiveService` or `Projects::UnarchiveService` if the `active` field is being changed. In that case, the contract checks of `Projects::ArchiveContract` or `Projects::UnarchiveContract` have already been run by `Projects::UpdateService`, so no need to use a contract once again. That's why [an EmptyContract is being used](https://github.com/opf/openproject/pull/11853/files#diff-da3e8ab4aa6c0709ac38d2a184dc26e40d7471a15726fd161ddd0c22ed17e86dR102-R105).

It's all in this commit: [`a28d1a05c0`](https://github.com/opf/openproject/pull/11853/commits/a28d1a05c0)

Also the fact that `allows_to?` was returning false, or that `Users::ProjectRoleCache` was returning no roles when project has `active == false` was annoying to be able to reuse the`Archive/Unarchive`-`Contract` in UpdateContract, because the instance being updated through the API could have the active flag being changed. So I introduced `Project#being_archived?` which checks if the project active state is being changed, so that things still work when this field is being changed. This allowed to remove the [`with_active_assumed` hacky method](https://github.com/opf/openproject/pull/11853/files#diff-969458ad4f527537566e3c9b1ccb9410dd0424d78637f65a8185bbb7d7977fb2L122-L129).

File: db/migrate/20221223084601_add_archive_project_permission_to_project_admin_role.rb
Comment (cbliard) [reply db/migrate/20221223084601_add_archive_project_permission_to_project_admin_role.rb]: I was not sure if that was a good idea or not. I'll remove it.

File: app/contracts/projects/base_contract.rb
Comment (cbliard) [reply app/contracts/projects/base_contract.rb:122]: I do not understand why the rest of the validations would run twice.

For instance the only things being checked by `Projects::ArchiveContract` are

```
    validate :validate_no_foreign_wp_references
    validate :validate_has_archive_project_permission
```

It won't run any other check.

Note that `Projects::ArchiveContract` inherits from `::BaseContract`, not `Projects::BaseContract`. That's maybe where the confusion is coming from. (and at first I forgot the `::`, and yes, the unit test was failing with a stack too deep error :) )

File: app/models/users/project_role_cache.rb
Comment (cbliard) [reply app/models/users/project_role_cache.rb]: good point :+1:

File: db/migrate/20221223084601_add_archive_project_permission_to_project_admin_role.rb
Comment (cbliard) [reply db/migrate/20221223084601_add_archive_project_permission_to_project_admin_role.rb]: Fixed

File: app/contracts/projects/archive_contract.rb
Comment (cbliard) [reply app/contracts/projects/archive_contract.rb]: I changed it to an early return:

```ruby
    def validate_can_archive_subprojects
      return if errors.any?
      return if user.allowed_to?(:archive_project, model.descendants)

      errors.add :base, :archive_permission_missing_on_subprojects
    end
```

Previous version would have been helpful if the subprojects names were added to the error message, but that's not the case.

Comment (cbliard) [reply app/contracts/projects/archive_contract.rb]: But it does not work because `user.allowed_to?(action, projects)` returns `false` if `projects` is an empty array.

I'll look into it tomorrow.

File: app/models/users/project_role_cache.rb
Comment (cbliard) [reply app/models/users/project_role_cache.rb]: Done

File: app/contracts/projects/archive_contract.rb
Comment (cbliard) [reply app/contracts/projects/archive_contract.rb]: Done in d91666d

File: config/initializers/permissions.rb
Comment (cbliard) [reply config/initializers/permissions.rb:42]: Parimal confirmed that it's ok if a user having only `archive_project` permission has to go to projects list to archive a project, and is not able to see the project settings page.
https://community.openproject.org/work_packages/3897/activity#activity-86

So I'll close this comment as "won't fix"

File: app/controllers/projects/archive_controller.rb
Comment (cbliard) [inline app/controllers/projects/archive_controller.rb:70]: @ulferts What is this `status` parameter? Is it really used?

Comment (ulferts) [reply app/controllers/projects/archive_controller.rb:70]: Good point. It looks to be a relic from when what is now the boolean property `active` was still the enum (with only two values) property `status`. My first urge would then be to simply change it to `active` but on the other hand, the code seems to work without that value, so it is worth to check whether `permit(:status)` can simply be removed.

Comment (cbliard) [reply app/controllers/projects/archive_controller.rb:70]: @ulferts It looks like it was used to filter the projects list to get active / archived projects ([39f53d2a00](https://github.com/opf/openproject/commit/39f53d2a00#diff-55c87ec2e453f92d3fe7bf0ec3369e8488b2b24621d8259e079ddc3875903c0aR305-R306)). And then on [4e5b9afef5](url), the status enum field was replaced by the active boolean field, and `params[:filters]` [started to be used](https://github.com/opf/openproject/commit/4e5b9afef5#diff-55c87ec2e453f92d3fe7bf0ec3369e8488b2b24621d8259e079ddc3875903c0aR249) instead (maybe not the commit where `params[:filters]` was first used, but that's the idea).

If I'm correct, we can probably remove this `project_path_with_status` and replace it with `projects_path`, right?

File: config/locales/en.yml
Comment (ulferts) [inline config/locales/en.yml]: I have to say that I don't find the last part of the error message particularly helpful ("Please contact your project administrator"). 

When the default seeding is used, the role called "Project admin" will get the role so a user having that role in the project and failing to archive is left wondering whom to contact. And even if the role might be named differently, the project administrator within the project the user just failed to archive might not have the necessary permissions in all the sub-projects. I guess it should be changed to "Please contact an administrator" because those are then the only ones who can reliably archive the project.

File: app/controllers/projects/archive_controller.rb
Comment (ulferts) [reply app/controllers/projects/archive_controller.rb:70]: Yes, looks like it.

File: app/contracts/projects/base_contract.rb
Comment (ulferts) [reply app/contracts/projects/base_contract.rb:122]: > Note that Projects::ArchiveContract inherits from ::BaseContract, not Projects::BaseContract

That was my mistake, yes. Sorry for the confusion.

File: app/contracts/projects/unarchive_contract.rb
Comment (ulferts) [reply app/contracts/projects/unarchive_contract.rb:30]: Admittedly you are right about this part. As said when we were talking, I misunderstood `::BaseContract` to mean `Projects::BaseContract` which is incorrect.

I have now looked through how the contract is currently used and your assessment seems to be right. But the way the Archive/Unarchive contracts are used currently is different than how we use them in other places where we typically do something like this:

1) change the attributes in a service
2) validate the changes via a contract
3) persist the changes triggered by the service (`model.save`) in case of confirmed validity

Here, the contracts are used only to check for permissions and for checking the current state of the projects and its descendants with the `active` attribute being unchanged. 

That definitely has the benefit of allowing the Contract to be callable from another one (e.g. `UpdateContract` in this case) without the hassle this would cause were the write protection in place.  

If the contract is misused, it will not offer any protection that might be expected by it, e.g.
```
project = Project.find('abc')
project.active = false
project.name = 'xyz'

Projects::ArchiveContract
  .new(project, User.admin.first)
  .valid?
```

would return `true` even though more is changed than what is necessary when archiving because the write protection does not exist. Admittedly, the write protection was too aggressive before because it also did not allow the `active` property to be written.

When arguing with the principle of least surprise in mind, I'd have preferred the contracts to keep inheriting from the `ModelContract`, adding `attribute :active` to the contracts and have the property changed in the service before the contract validates. But that would clash seriously with the way the contracts are now used by the `UpdateContract` which has merits of its own.

TL;DR In conclusion I agree that the approach you took makes sense if only because of the way the contracts are used in this specific case.

**Review by ulferts — CHANGES_REQUESTED** [review]
With my misconception about the `::BaseService` being cleared up, I only have a tiny thing left about the error message that I don't think optimal.

File: app/contracts/projects/unarchive_contract.rb
Comment (cbliard) [reply app/contracts/projects/unarchive_contract.rb:30]: I understand your concerns.

On the other hand, as you said, the previous version of ArchiveContract was too agressive:
on `dev`

```ruby
project = Project.last
project.active = false
Projects::ArchiveContract.new(project, User.admin.first).valid?
=> false
```

It looks more like a `ArchivableContract`.

Is there some kind of guards we could use to make sure it is not used in an unintended way?

File: config/locales/en.yml
Comment (cbliard) [reply config/locales/en.yml]: That's how it has been specified in the original ticket. Your remark makes sense.

I'll update it as you proposed and notify @psatyal

Comment (cbliard) [reply config/locales/en.yml]: Done in b664cf6d

File: app/contracts/projects/unarchive_contract.rb
Comment (cbliard) [reply app/contracts/projects/unarchive_contract.rb:30]: An alternative would be to just delete `Projects::ArchiveContract` and `Projects::UnarchiveContract`, and put all the logic in `Projects::UpdateContract`, and this contract would be used for all of `Projects::ArchiveService`, `Projects::UnarchiveService`, and `Projects::UpdateService`.

On the other hand, `Projects::ArchiveService`, `Projects::UnarchiveService` are overriding `#persist` to only update the `active` field.

Comment (ulferts) [reply app/contracts/projects/unarchive_contract.rb:30]: > Is there some kind of guards we could use to make sure it is not used in an unintended way?

Can't think of one currently.

**Review by ulferts — APPROVED** [review]

File: app/controllers/projects/archive_controller.rb
Comment (cbliard) [reply app/controllers/projects/archive_controller.rb:70]: I'll refactor this once merged.

File: config/locales/en.yml
Comment (cbliard) [reply config/locales/en.yml]: And [Parimal approved](https://community.openproject.org/projects/openproject/work_packages/3897/activity#activity-89).

**Pull request merged**
